corehq/apps/data_cleaning/models.py

```diff
@@ -269,6 +269,30 @@ def get_num_selected_records_in_queryset(self):
 
         return num_selected_records
 
+    def _apply_change_to_selected_doc_ids(self, change, doc_ids):
+        """
+        Apply a change to the selected records with the given doc_ids.
+        :param change: BulkEditChange
+        :param doc_ids: list of doc ids
+        """
+        selected_records = BulkEditRecord.get_selected_records_with_ids(self, doc_ids)
+        change.records.add(*selected_records)
+
+    def apply_change_to_selected_records_in_queryset(self, change):
+        """
+        Apply a change to the selected records in the current queryset.
+        :param change: BulkEditChange
+        """
+        if self.has_any_filtering:
+            self._apply_operation_on_queryset(
+                lambda doc_ids: self._apply_change_to_selected_doc_ids(change, doc_ids)
+            )
+        else:
+            # If there are no filters, we can just apply the change to all selected records
+            # this will be a faster operation for larger data sets
+            selected_records = self.records.filter(is_selected=True)
+            change.records.add(*selected_records)
```

Comment:
I'm not sure how django treats `add` if `selected_records` is empty. Is it possible that `selected_records` can be empty here, and if so, will this safely be a no-op?

Reply:
no it is not possible that `selected_records` will be empty here. the validation for that is in the UI

Reply:
We want this code to be as resilient as possible to changes outside of its scope. If its currently the case that the UI prevents `selected_records` from being empty, a developer could come along at a later date and change that UI (or re-use this code with another UI), where that is no longer true. Again, I don't know whether Django would gracefully handle `selected_records` being empty here, but if it doesn't, checking for empty makes the code more stable